### PR TITLE
PG-2287 - Add additional testcase for pg_tde_rewind to cover edge case

### PR DESCRIPTION
**Pull request by mohitj1988**

### CODE REVIEW

File: postgresql/automation/tests/pg_tde_rewind_extended.sh [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+    FOR VALUES FROM ('2024-01-01') TO ('2024-06-01');
+  "
+  $PSQL -p $REPLICA_PORT -c "
+    INSERT INTO part_test SELECT g, '2024-02-01' FROM generate_series(1,10000) g;
+  "
+
+  # UNLOGGED + TEMP tables
+  $PSQL -p $REPLICA_PORT -c "CREATE UNLOGGED TABLE u1(id INT);"
+  $PSQL -p $REPLICA_PORT -c "INSERT INTO u1 SELECT generate_series(1,10000);"
+  $PSQL -p $REPLICA_PORT -c "CREATE TEMP TABLE temp1(id INT); INSERT INTO temp1 VALUES (1),(2);"
```
Comment (shahidullah79) [inline postgresql/automation/tests/pg_tde_rewind_extended.sh:187]: Temp tables are session-scoped — useless across connections — and combining them in one -c call is fragile.
$PSQL -p $REPLICA_PORT -c "CREATE TEMP TABLE temp1(id INT);"
$PSQL -p $REPLICA_PORT -c "INSERT INTO temp1 VALUES (1),(2);"

File: postgresql/automation/tests/pg_rewind_extended.sh [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+    --time=30 run
+
+  ###########################################
+  # Crash simulation
+  ###########################################
+  if [ "$CRASH_MODE" -eq 1 ]; then
+    echo "Simulating crash on replica"
+
+    kill -9 $(head -1 $REPLICA_DATA/postmaster.pid)
+    sleep 2
```
Comment (shahidullah79) [inline postgresql/automation/tests/pg_rewind_extended.sh]: crash_pg from common.sh additionally waits for child processes and cleans the socket file. Raw kill -9 + sleep 2 can leave stale sockets causing start_pg to fail.

crash_pg $REPLICA_DATA

File: postgresql/automation/tests/pg_rewind_extended.sh [hunk 3]
```diff
@@ -0,0 +1,298 @@
+#!/bin/bash
+
+#############################################
+# CONFIG
+#############################################
+KEYFILE="/tmp/keyring.file"
```
Comment (shahidullah79) [inline postgresql/automation/tests/pg_rewind_extended.sh]: better to have keyring.file in some other place instead of /tmp. may be
KEYFILE="$RUN_DIR/keyring.file"
is better

File: postgresql/automation/tests/pg_rewind_extended.sh [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+KEY_ROTATION=${KEY_ROTATION:-1}
+CRASH_MODE=${CRASH_MODE:-1}
+TABLESPACE_TEST=${TABLESPACE_TEST:-1}
+TABLESPACE_FILE="/tmp/ts1"
+TABLESPACE_FILE_REPL="/tmp/ts1_replica"
```
Comment (shahidullah79) [inline postgresql/automation/tests/pg_rewind_extended.sh]: similarly TABLESPACE_FILE uses /tmp instead of $RUN_DIR

File: postgresql/automation/tests/pg_rewind_extended.sh [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+PG_REWIND="$INSTALL_DIR/bin/pg_tde_rewind"
+PSQL="$INSTALL_DIR/bin/psql"
+
+#############################################
+# CLEANUP
+#############################################
+echo "Cleaning environment"
+old_server_cleanup $PRIMARY_DATA
+old_server_cleanup $REPLICA_DATA
+rm -rf $ARCHIVE_DIR $KEYFILE $TABLESPACE_FILE $TABLESPACE_FILE_REPL || true
```
Comment (shahidullah79) [inline postgresql/automation/tests/pg_rewind_extended.sh]: Unquoted path variables throughout. 
rm -rf "$ARCHIVE_DIR" "$KEYFILE" "$TABLESPACE_FILE" "$TABLESPACE_FILE_REPL" || true
cp -R "$PRIMARY_DATA/pg_tde" "$REPLICA_DATA/"

File: postgresql/automation/tests/pg_tde_rewind_extended.sh [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+
+echo "host replication all 127.0.0.1/32 trust" >> $PRIMARY_DATA/pg_hba.conf
+
+start_pg $PRIMARY_DATA $PRIMARY_PORT
+
+$PSQL -p $PRIMARY_PORT -d postgres -c "CREATE EXTENSION pg_tde;"
+$PSQL -p $PRIMARY_PORT -d postgres -c "SELECT pg_tde_add_global_key_provider_file('file_provider','$KEYFILE');"
+$PSQL -p $PRIMARY_PORT -d postgres -c "SELECT pg_tde_create_key_using_global_key_provider('key1','file_provider');"
+$PSQL -p $PRIMARY_PORT -d postgres -c "SELECT pg_tde_set_default_key_using_global_key_provider('key1','file_provider');"
+
```
Comment (shahidullah79) [inline postgresql/automation/tests/pg_tde_rewind_extended.sh:63]: psql exits 0 even on SQL errors. A failed CREATE EXTENSION silently continues.
Add --variable ON_ERROR_STOP=1 to critical setup calls:
$PSQL -p $PRIMARY_PORT -d postgres --variable ON_ERROR_STOP=1 -c "CREATE EXTENSION pg_tde;"

File: postgresql/automation/tests/pg_rewind_with_checkpoint.sh [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+# Step 9: Verify data
+#######################################
+echo "Querying table after rewind"
+
+$PSQL -p $PRIMARY_PORT -d postgres -c "SELECT count(*) FROM sbtest1;"
+$PSQL -p $PRIMARY_PORT -d postgres -c "SELECT count(*) FROM sbtest2;"
+$PSQL -p $PRIMARY_PORT -d postgres -c "SELECT count(*) FROM sbtest10;"
+$PSQL -p $PRIMARY_PORT -d postgres -c "SELECT count(*) FROM sbtest15;"
+$PSQL -p $PRIMARY_PORT -d postgres -c "SELECT count(*) FROM sbtest20;"
+$PSQL -p $PRIMARY_PORT -d postgres -c "SELECT count(*) FROM sbtest25;"
```
Comment (shahidullah79) [inline postgresql/automation/tests/pg_rewind_with_checkpoint.sh]: These queries print output but never check if counts are correct. The test passes even if data is missing or wrong.

add something like this.
COUNT=$(PSQL -p $PRIMARY_PORT -d postgres -At -c "SELECT count(*) FROM sbtest1;")
if [ "$COUNT" -lt 1000 ]; then
  echo "FAIL: sbtest1 count $COUNT < 1000"
  exit 1
fi

File: postgresql/automation/tests/pg_rewind_with_checkpoint.sh [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+echo "listen_addresses = '*'" >> $REPLICA_DATA/postgresql.conf
+echo "logging_collector = on" >> $REPLICA_DATA/postgresql.conf
+echo "log_directory = '$REPLICA_DATA'" >> $REPLICA_DATA/postgresql.conf
+echo "log_filename = 'server.log'" >> $REPLICA_DATA/postgresql.conf
+echo "log_statement = 'all'" >> $REPLICA_DATA/postgresql.conf
+echo "default_table_access_method = 'tde_heap'" >> $REPLICA_DATA/postgresql.conf
+echo "max_wal_senders = 5" >> $REPLICA_DATA/postgresql.conf
+echo "io_method = '$IO_METHOD'" >> $REPLICA_DATA/postgresql.conf
+echo "shared_preload_libraries = 'pg_tde'" >> $REPLICA_DATA/postgresql.conf
+echo "max_wal_senders=10" >> $REPLICA_DATA/postgresql.conf
```
Comment (shahidullah79) [inline postgresql/automation/tests/pg_rewind_with_checkpoint.sh]: Duplicate value
echo "max_wal_senders = 5" >> $REPLICA_DATA/postgresql.conf   # line 61
...
echo "max_wal_senders=10" >> $REPLICA_DATA/postgresql.conf    # line 64

File: postgresql/automation/tests/pg_rewind_with_checkpoint.sh [hunk 9]
```diff
@@ -0,0 +1,137 @@
+#!/bin/bash
+
+# Binaries
+PG_CTL="$INSTALL_DIR/bin/pg_ctl"
+PG_BASEBACKUP="$INSTALL_DIR/bin/pg_tde_basebackup"
+PG_REWIND="$INSTALL_DIR/bin/pg_tde_rewind"
+PSQL="$INSTALL_DIR/bin/psql"
+INITDB="$INSTALL_DIR/bin/initdb"
```
Comment (shahidullah79) [inline postgresql/automation/tests/pg_rewind_with_checkpoint.sh]: Never used anywhere in the file.

File: postgresql/automation/tests/pg_rewind_with_checkpoint.sh [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+
+start_pg $REPLICA_DATA $REPLICA_PORT
+
+
+#######################################
+# Step 3: Create table on primary
+#######################################
+echo "Creating table and inserting data on primary"
+$SYSBENCH /usr/share/sysbench/oltp_insert.lua --pgsql-user=`whoami` --pgsql-db=postgres --db-driver=pgsql --pgsql-port=$PRIMARY_PORT --threads=5 --tables=100 --table-size=1000 prepare
```
Comment (shahidullah79) [inline postgresql/automation/tests/pg_rewind_with_checkpoint.sh]: better to use
--pgsql-user=$(whoami)

File: postgresql/automation/tests/pg_rewind_with_checkpoint.sh [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+$PG_CTL -D $REPLICA_DATA promote
+sleep 3
+
+#######################################
+# Step 5: Diverging writes on replica
+#######################################
+echo "Inserting more data on promoted replica"
+
+$PSQL -p $REPLICA_PORT -d postgres -c "INSERT INTO t1 VALUES (4),(5),(6);"
+$SYSBENCH /usr/share/sysbench/oltp_read_write.lua --pgsql-user=`whoami` --pgsql-db=postgres --db-driver=pgsql --pgsql-port=$REPLICA_PORT --threads=5 --tables=100 --time=60 --report-interval=10 run
```
Comment (shahidullah79) [inline postgresql/automation/tests/pg_rewind_with_checkpoint.sh]: same here 
--pgsql-user=$(whoami) instead of --pgsql-user=`whoami`

File: postgresql/automation/tests/pg_rewind_with_checkpoint.sh [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+echo "logging_collector = on" >> $REPLICA_DATA/postgresql.conf
+echo "log_directory = '$REPLICA_DATA'" >> $REPLICA_DATA/postgresql.conf
+echo "log_filename = 'server.log'" >> $REPLICA_DATA/postgresql.conf
+echo "log_statement = 'all'" >> $REPLICA_DATA/postgresql.conf
+echo "default_table_access_method = 'tde_heap'" >> $REPLICA_DATA/postgresql.conf
+echo "max_wal_senders = 5" >> $REPLICA_DATA/postgresql.conf
+echo "io_method = '$IO_METHOD'" >> $REPLICA_DATA/postgresql.conf
+echo "shared_preload_libraries = 'pg_tde'" >> $REPLICA_DATA/postgresql.conf
+echo "max_wal_senders=10" >> $REPLICA_DATA/postgresql.conf
+echo "restore_command='cp $ARCHIVE_DIR/%f %p'" >> $REPLICA_DATA/postgresql.conf
```
Comment (shahidullah79) [inline postgresql/automation/tests/pg_rewind_with_checkpoint.sh]: 16 separate echo >> calls instead of a single cat > ... <<EOF block. Harder to read and audit.

Fix: Replace with a heredoc like pg_rewind_extended.sh does:
cat > $REPLICA_DATA/postgresql.conf <<EOF
port=$REPLICA_PORT
unix_socket_directories='$RUN_DIR'
...
EOF

File: postgresql/automation/tests/pg_rewind_extended.sh [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+  rewind_and_start
+
+  ###########################################
+  # Validation (improved)
+  ###########################################
+  $PSQL -p $PRIMARY_PORT -c "SELECT count(*) FROM t1;"
+
+  echo "Deep validation"
+  $PSQL -p $PRIMARY_PORT -c "SET enable_seqscan=off;"
+  $PSQL -p $PRIMARY_PORT -c "SELECT * FROM t1 ORDER BY id LIMIT 10;"
```
Comment (shahidullah79) [inline postgresql/automation/tests/pg_rewind_extended.sh]: SET enable_seqscan=off has no effect. 
$PSQL -p $PRIMARY_PORT -c "SET enable_seqscan=off;"
$PSQL -p $PRIMARY_PORT -c "SELECT * FROM t1 ORDER BY id LIMIT 10;"
These are two separate connections. The SET is discarded immediately.
Fix: 
$PSQL -p $PRIMARY_PORT -c "SET enable_seqscan=off; SELECT * FROM t1 ORDER BY id LIMIT 10;"

**Review by dAdAbird — COMMENTED** [review]
A great starting point.
Things that I'd add:
1. Now pg_tde_rewind run one with the local source `--source-pgdata=$REPLICA_DATA `. We have to also test the remote source, like `--source-server="port=5433 host=localhost"`
2. Recreate not only principal keys but also create and change providers.
3. Restart primary / replica before and after promotion to generate new WAL keys

And I'd separate all those to different scripts because overlapping of actions may "fix" things.
For example, here is my script to trigger scenario when encrypted file on target remains unchanged by rewind:
```sh
create_primary
reate_replica

psql -p $PRIMARY_PORT -c "
CREATE TABLE t3 (id INTEGER GENERATED ALWAYS AS IDENTITY PRIMARY KEY, f1 TEXT) USING tde_heap;
INSERT INTO t3 (f1) SELECT repeat('abcdeF', 10) FROM generate_series(1, 1000);
"

pg_ctl restart -D $PGDATA_PRI

promote_replica

psql -p $REPL_PORT -c "
INSERT INTO t3 (f1) SELECT repeat('abcdeF', 10) FROM generate_series(1, 10);
checkpoint;
"

do_rewind
check_results
```

This makes rewind to keep an index relation on target. And this is the case that should be handled in terms of keys.
But running some other action of table may cause rewind to copy the index relation from source and that won't trigger the case.

File: postgresql/automation/tests/pg_rewind_extended.sh (hunk 3 above)
Comment (mohitj1988) [reply postgresql/automation/tests/pg_rewind_extended.sh]: Fixed

File: postgresql/automation/tests/pg_rewind_extended.sh (hunk 4 above)
Comment (mohitj1988) [reply postgresql/automation/tests/pg_rewind_extended.sh]: Fixed

File: postgresql/automation/tests/pg_rewind_extended.sh (hunk 5 above)
Comment (mohitj1988) [reply postgresql/automation/tests/pg_rewind_extended.sh]: Fixed

File: postgresql/automation/tests/pg_rewind_extended.sh (hunk 2 above)
Comment (mohitj1988) [reply postgresql/automation/tests/pg_rewind_extended.sh]: Fixed

File: postgresql/automation/tests/pg_rewind_extended.sh (hunk 13 above)
Comment (mohitj1988) [reply postgresql/automation/tests/pg_rewind_extended.sh]: Fixed

File: postgresql/automation/tests/pg_rewind_with_checkpoint.sh (hunk 9 above)
Comment (mohitj1988) [reply postgresql/automation/tests/pg_rewind_with_checkpoint.sh]: Removed

File: postgresql/automation/tests/pg_rewind_with_checkpoint.sh (hunk 8 above)
Comment (mohitj1988) [reply postgresql/automation/tests/pg_rewind_with_checkpoint.sh]: Fixed

File: postgresql/automation/tests/pg_rewind_with_checkpoint.sh (hunk 12 above)
Comment (mohitj1988) [reply postgresql/automation/tests/pg_rewind_with_checkpoint.sh]: Fixed

File: postgresql/automation/tests/pg_rewind_with_checkpoint.sh (hunk 10 above)
Comment (mohitj1988) [reply postgresql/automation/tests/pg_rewind_with_checkpoint.sh]: Fixed

File: postgresql/automation/tests/pg_rewind_with_checkpoint.sh (hunk 11 above)
Comment (mohitj1988) [reply postgresql/automation/tests/pg_rewind_with_checkpoint.sh]: Fixed

File: postgresql/automation/tests/pg_rewind_with_checkpoint.sh (hunk 7 above)
Comment (mohitj1988) [reply postgresql/automation/tests/pg_rewind_with_checkpoint.sh]: We are using sysbench to randomly insert rows. We cannot be sure if rows are always less than 1000. It can be more than 1000 also.

The idea of querying the table was to ensure there is no table corruption introduced. For deterministic results I'll add a check for count > 0

**Comment by mohitj1988** [comment]
@dAdAbird Hi,
1. Now pg_tde_rewind run one with the local source --source-pgdata=$REPLICA_DATA . We have to also test the remote source, like --source-server="port=5433 host=localhost"

Regarding this, we have several pg_tde_rewind tests that are testing this. Ex
https://github.com/Percona-QA/percona-qa/blob/master/postgresql/automation/tests/pg_tde_rewind_test.sh
https://github.com/Percona-QA/percona-qa/blob/master/postgresql/automation/tests/pg_tde_rewind_loop_test.sh

2. Recreate not only principal keys but also create and change providers.
Sure, it is a good idea to create/change providers as well

3. Restart primary / replica before and after promotion to generate new WAL keys
Sure, I'll add

File: postgresql/automation/tests/pg_tde_rewind_extended.sh [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+  restart_pg $REPLICA_DATA $REPLICA_PORT
+
+  ###########################################
+  # Promote replica
+  ###########################################
+  $PG_CTL -D $REPLICA_DATA promote
+  sleep 3
+
+  restart_pg $PRIMARY_DATA $PRIMARY_PORT
+  restart_pg $REPLICA_DATA $REPLICA_PORT
```
Comment (dAdAbird) [inline postgresql/automation/tests/pg_tde_rewind_extended.sh:132]: Here and L122-123, I think it make sense at introduce randomness and asymmetry with these restarts. Meaning, that some tests w/o restarts at all, and some do them partially or all of them. Therefore we greater variety and test different scenarios: 1) when no extra WAL keys, 2) or those only on 3) source or target, 4) or on both.

**Review by dAdAbird — COMMENTED** [review]
1. With pg_rewind in particular I found that introduction of extra steps (new queries, restarts etc) might change the outcome by overshadowing ("fixing") issues. For example having only Insert + checkpoint might trigger partial write of the relation and exposes issues there. But by adding extra requests, we might trigger rewind to rewrite relation in full, hence taking another execution part. I dunno what to do with that. Maybe, again, add randomness to if some queries are run or not.
2. Also would be nice to have asymmetry - when some tables/databases? are added on both servers after replica promotion. Any new tables on target would be wiped out but all those are from source should exists after the rewind.
3. We should also test running pg_tde_rewind on cluster w/o "pg_tde" lib loaded.

I'm not saying that all of this should be necessary in this PR, but we definitely need to test that

File: postgresql/automation/tests/pg_tde_rewind_extended.sh (hunk 14 above)
Comment (mohitj1988) [reply postgresql/automation/tests/pg_tde_rewind_extended.sh:132]: Added a new test pg_tde_rewind_randomized.sh to cover these scenarios

**Review by shahidullah79 — APPROVED** [review]
LGTM
